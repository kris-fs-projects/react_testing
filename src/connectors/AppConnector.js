import { connect } from "react-redux";
import { createGetArtistListAsyncAction } from "../actions/artistActionCreators";
import { createGetRecordListAsyncAction } from "../actions/recordActionCreators";
import App from "../App";

const mapStateToProps = (store) => ({
  selectedArtist: store.artistsReducer.selectedArtist,
  selectedRecord: store.recordsReducer.selectedRecord
});

const mapDispatchToProps = (dispatch) => ({
  fetchArtistsList: () => dispatch(createGetArtistListAsyncAction()),
  fetchRecordsList: () => dispatch(createGetRecordListAsyncAction())
});

export const AppConnector = connect(mapStateToProps, mapDispatchToProps)(App);

/* 
    <App 
        selectedArtist={}
        selectedRecord={}
    />
*/