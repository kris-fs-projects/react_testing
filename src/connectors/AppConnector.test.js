import { render, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import { AppConnector } from "./AppConnector";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";

const middleware = [thunk];
const mockStore = configureStore(middleware);

// see https://www.robinwieruch.de/react-connected-component-test/
describe('App Connector', () => {

  let store;
  let mockData = {
    artistsReducer: {
      artists: [
        {
          id: '1',
          name: 'The Tragically Hip'
        },
        {
          id: "2",
          name: "The Watchmen"
        },
      ],
      selectedArtist: {
        id: '1',
        name: 'The Tragically Hip'
      }
    },
    recordsReducer: {
      records: [
        {
          id: '1',
          title: 'Up To Here',
          artistId: '1',
          year: '1989'
        },
        {
          id: '2',
          title: 'Saskadelphia',
          artistId: '1',
          year: '2021'
        },
        {
          id: "3",
          title: "McLaren Furnace Room",
          artistId: "2",
          year: "1992"
        },
      ],
      filteredRecords: [
        {
          id: '1',
          title: 'Up To Here',
          artistId: '1',
          year: '1989'
        },
        {
          id: '2',
          title: 'Saskadelphia',
          artistId: '1',
          year: '2021'
        },
      ],
      selectedRecord: {
        id: '2',
        title: 'Saskadelphia',
        artistId: '1',
        year: '2021'
      }
    }
  }

  beforeEach(() => {
    store = mockStore(mockData);
  });

  it('should render with the state from the store', () => {
    // arrange
    const { asFragment } = render(<Provider store={store}>
      <AppConnector />
    </Provider>);

    // act

    // assert
    expect(asFragment()).toMatchSnapshot();
  });

  it('should display the artists, the records for the selected artist and the details of the selected record', () => {
    // arrange
    render(<Provider store={store}>
      <AppConnector />
    </Provider>);

    // act

    //assert
    const artist1 = screen.getByText(mockData.artistsReducer.artists[0].name);
    const artist2 = screen.getByText(mockData.artistsReducer.artists[1].name);
    const selectedArtistRecordsHeading = screen.getByText(`${mockData.artistsReducer.selectedArtist.name} Records`);
    const record1 = screen.getByText(mockData.recordsReducer.filteredRecords[0].title);
    const record2 = screen.getAllByText(mockData.recordsReducer.filteredRecords[1].title)[0];
    const selectedRecordHeading = screen.getByText(`${mockData.recordsReducer.selectedRecord.title} 411`);
    const selectedRecordYear = screen.getByText(mockData.recordsReducer.selectedRecord.year);

    expect(artist1).toBeInTheDocument();
    expect(artist2).toBeInTheDocument();
    expect(selectedArtistRecordsHeading).toBeInTheDocument();
    expect(record1).toBeInTheDocument();
    expect(record2).toBeInTheDocument();
    expect(selectedRecordHeading).toBeInTheDocument();
    expect(selectedRecordYear).toBeInTheDocument();

    expect(artist1).toHaveClass('selected');
    expect(artist2).toHaveClass('normal');

    expect(record1).toHaveClass('normal');
    expect(record2).toHaveClass('selected');
  });

});