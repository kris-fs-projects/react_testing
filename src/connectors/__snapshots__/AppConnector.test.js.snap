// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Connector should render with the state from the store 1`] = `
<DocumentFragment>
  <div
    class="rolodexContainer"
  >
    <header>
      <h1>
        my record rolodex
      </h1>
    </header>
    <div
      class="fullHeight"
    >
      <div
        class="artistsBox"
      >
        <div
          class="columnWrapper"
        >
          <div
            class="column"
          >
            <h2
              class="columnHeader"
            >
              Artists
            </h2>
            <div
              class="columnChildren"
            >
              <input
                placeholder="type a new artist name and hit enter to save it"
                style="width: 100%; margin: 2px; box-sizing: border-box;"
              />
              <ul>
                <li
                  class="listItem selected"
                >
                  The Tragically Hip
                </li>
                <li
                  class="listItem normal"
                >
                  The Watchmen
                </li>
              </ul>
            </div>
          </div>
        </div>
      </div>
      <div
        class="recordsBox"
      >
        <div
          class="columnWrapper"
        >
          <div
            class="column"
          >
            <h2
              class="columnHeader"
            >
              The Tragically Hip Records
            </h2>
            <div
              class="columnChildren"
            >
              <div
                class="formWrapper"
              >
                <input
                  class="formInput"
                  placeholder="album title"
                  style="width: 50%;"
                />
                <input
                  class="formInput"
                  placeholder="year"
                  style="width: 20%;"
                />
                <button
                  style="width: 15%;"
                >
                  Save
                </button>
              </div>
              <ul>
                <li
                  class="listItem normal"
                >
                  Up To Here
                </li>
                <li
                  class="listItem selected"
                >
                  Saskadelphia
                </li>
              </ul>
            </div>
          </div>
        </div>
      </div>
      <div
        class="detailsBox"
      >
        <div
          class="columnWrapper"
        >
          <div
            class="column"
          >
            <h2
              class="columnHeader"
            >
              Saskadelphia 411
            </h2>
            <div
              class="columnChildren"
            >
              <div
                class="selected"
              >
                <p>
                  Saskadelphia
                </p>
                <p>
                  2021
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <footer>
      Â© 2022 Your Friendly Neighbourhood Record Store
    </footer>
  </div>
</DocumentFragment>
`;
